File: C/p003.c
/*
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
*/

#include <stdio.h>
#include <math.h>

int is_multiple(unsigned long long, unsigned long long);
int is_prime(unsigned long long);
unsigned long long largest_prime_factor(unsigned long long);

int is_multiple(unsigned long long num, unsigned long long divisor) {
    return !(num % divisor);
}

int is_prime(unsigned long long num) {
    if (num == 1) {
        return 0;
    }
    if (num == 2) {
        return 1;
    }
    if (is_multiple(num, 2)) {
        return 0;
    }

    double limit = sqrt((double) num);
    unsigned long long i;

    for (i = 3; i <= limit; i += 2) {
        if (is_multiple(num, i)) {
            return 0;
        }
    }
    return 1;
}

unsigned long long largest_prime_factor(unsigned long long num) {
    unsigned long long limit = num;
    unsigned long long i = 2;

    while (i < num) {
        if (i == num) {
            return i;
        }
        if (is_multiple(num, i)) {
            num /= i;
            continue;
        }

        if (i == 2) {
            i++;
        } else {
            i += 2;
        }
        for (; i < num && !is_prime(i); i += 2)
            ;
    }
}

int main() {
    printf("%llu\n", largest_prime_factor(600851475143LL));
    return 0;
}

File: Python/p003.py
import math

prime_number_list = [2,3]

def primes(limit=None):
    global prime_number_list

    for prime in prime_number_list:
        if limit and prime > limit: break
        yield prime

    num = prime_number_list[-1]
    while True:
        num += 2

        if limit and num > limit: break

        if any( num % prime == 0 for prime in prime_number_list ): continue

        prime_number_list.append(num)
        yield num

def primes_old(limit):
    prime_list = [2,3]

    yield 2
    yield 3
    num = 3

    while num + 2 < limit:
        num += 2

        if any( num % prime == 0 for prime in prime_list ): continue

        prime_list.append(num)
        yield num

def prime_list(limit):
    return list(primes(limit))

def highest_prime_factor(number):
    
    limit = math.sqrt(number)

    divisor = 2

    while divisor < limit and number != divisor:
        if number % divisor == 0:
            number /= divisor
            continue
        divisor += 1 if divisor == 2 else 2
    
    return number

if __name__ == "__main__":
    print highest_prime_factor(600851475143)
